src/Datadog.Trace.ClrProfiler.Managed/ScopeFactory.cs

```diff
@@ -191,33 +192,30 @@ public static Scope CreateDbCommandScope(Tracer tracer, IDbCommand command)
             return scope;
         }
 
-        public static string GetDbType(string commandTypeName)
+        public static string GetDbType(string namespaceName, string commandTypeName)
```

Comment:
Can you add a benchmark to this change? This method is used in the integration hotpath, and the codegen is already bigger: 

![image](https://user-images.githubusercontent.com/69803/116090168-df3b0800-a6a3-11eb-93cd-a44b5916a370.png)

Reply:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBkBLAOwEcBuGjWIBmJqQYBhBgG8aDeQvkim5JCsYB5ADYATAOIwMAEWAAVAJ4AHGAApmjSPnzYuOi9YBy2fDACUcxQVZakDQhlwAdw4MMAALBhtHZ1d3GC8ffxCwxWDs7LA8GAYAIgBlHi0JCCcXHWKQALy84gB2EtwKuFwYKAA3HuKBLKbFAu6Sj0sAcw7K6uS6huGR0NaSywhcDCnYXEHGlYYxouKAWXNyuZrXeoPDhTXi/HNZ/eWV45KNKGwwLRgqtdFnd7kw2sVoL9/m9QUdCiVLtEAfNard3iNPmUAIqcDDIoFo2GKR6zJEw0GYgCSXDxUDAMEsGGwwH+JkBC0JFPhxQAClAIAAzDj/HRslE3JZEhQAemlDB0EC4AHIMEdYNg8eFLC5cAwBdAGBhYjBuiD7msuABXLRaIawnQwAXYa1YM2HSBcLbqI7inSlS0CoUIBgAXhK7NRQzdK1lDFg+Ag/XDvuK4QDQYYHAFDEsuxgNOjIzWSVqqXSMBYAFFXLgAOrRWKJX3+wMcBCZKWHAD8PqBZe8Ff9wHsNjoaF7C37PnY+amRoYcAntRbQZnXDnsV8LFMEDYEAiPWpvWwUA4LgwNg7nZGICXKSsaQH293+8PXGPp/Pl7tIwAvm7/2WN1lGYNR7AYAxjDMB87HIRguAHXBtXpctx3Akt708Acr1CXIMUVL10ObdM21DZMCSjdFQg/OMTRdUNC1CDC3AfctwiiGJYkYwI8KlMoKgjG5QwAPnaTpuj6AY0G47JikmGYBJTET1k2bZdmKaSqMOM4LkUgllKeF4Kg0mSwn4rQkUEuoDNJPETK0lZsVxfEORsioyU069impWl6UZZlWWAKzUxDUSrRtTzO15fkhRFMV9NChhwq0SKpQAfQYCJjS4JLEOQx9pzYWd5xDMNGAAMnKu8WKwnxQzDZiV1I0LTKaZipwrHc9wPKAjxPM8aUvVKiQyrL82qjr6uqprgxahzQQQnwkN+ArB0sCyLyVFglS3Ng8AvLcutfXr336r9fGG2FRuyibWKfasdDrBsmyBGbfBE1q8nau7pyHEcx1u2qKyK9d50XRqSIQNcN0Ol8er6z9Bouz7FAyxLvqB59urfD8BoOz7f0olY1l2F0f1CQDfyAA=

Reply:
okay, will do.

Reply:
It seems to me the new code preforms similarly for the command cases, where non standard names are used, the new way of doing things is slightly slower, but in a way that seems acceptable to me.

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19042
Intel Core i7-10710U CPU 1.10GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=5.0.101
  [Host]     : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
  DefaultJob : .NET Core 3.1.10 (CoreCLR 4.700.20.51601, CoreFX 4.700.20.51901), X64 RyuJIT
```

|       Method |        namespaceName |   commandTypeName |       Mean |     Error |    StdDev |     Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |--------------------- |------------------ |-----------:|----------:|----------:|-----------:|-------:|------:|------:|----------:|
|    GetDbType |                      |           Command |  11.875 ns | 0.0584 ns | 0.0546 ns |  11.876 ns |      - |     - |     - |         - |
| OldGetDbType |                    ? |           Command |  15.267 ns | 0.1851 ns | 0.1641 ns |  15.186 ns |      - |     - |     - |         - |
|    GetDbType |            Custom.DB |           Command |  48.880 ns | 0.1343 ns | 0.1256 ns |  48.859 ns | 0.0007 |     - |     - |      64 B |
| OldGetDbType |                    ? |           Command |  14.645 ns | 0.1822 ns | 0.1705 ns |  14.604 ns |      - |     - |     - |         - |
|    GetDbType |                      |           Example | 113.167 ns | 0.4576 ns | 0.4057 ns | 113.206 ns | 0.0004 |     - |     - |      40 B |
| OldGetDbType |                    ? |           Example |  96.544 ns | 0.4771 ns | 0.3984 ns |  96.674 ns | 0.0004 |     - |     - |      40 B |
|    GetDbType |                      |    ExampleCommand | 134.179 ns | 0.3551 ns | 0.3148 ns | 134.218 ns | 0.0007 |     - |     - |      80 B |
| OldGetDbType |                    ? |    ExampleCommand | 120.796 ns | 2.4545 ns | 2.5206 ns | 120.110 ns | 0.0007 |     - |     - |      80 B |
|    GetDbType |                      | ProfiledDbCommand |  11.059 ns | 0.0725 ns | 0.0605 ns |  11.050 ns |      - |     - |     - |         - |
| OldGetDbType |                    ? | ProfiledDbCommand |  11.923 ns | 0.3075 ns | 0.7426 ns |  11.686 ns |      - |     - |     - |         - |
|    GetDbType | MySql(...)lient [22] |      MySqlCommand |  10.679 ns | 0.0465 ns | 0.0413 ns |  10.665 ns |      - |     - |     - |         - |
| OldGetDbType |                    ? |      MySqlCommand |  10.402 ns | 0.2500 ns | 0.2088 ns |  10.343 ns |      - |     - |     - |         - |
|    GetDbType |               Npgsql |     NpgsqlCommand |   9.808 ns | 0.2537 ns | 0.1981 ns |   9.730 ns |      - |     - |     - |         - |
| OldGetDbType |                    ? |     NpgsqlCommand |   9.589 ns | 0.0411 ns | 0.0365 ns |   9.595 ns |      - |     - |     - |         - |
|    GetDbType | Syste(...)lient [21] |        SqlCommand |   9.218 ns | 0.0542 ns | 0.1082 ns |   9.182 ns |      - |     - |     - |         - |
| OldGetDbType |                    ? |        SqlCommand |   8.893 ns | 0.0641 ns | 0.0535 ns |   8.904 ns |      - |     - |     - |         - |

Benchmark code was:

```
        [Benchmark]
        [Arguments("SqlCommand")]
        [Arguments("MySqlCommand")]
        [Arguments("NpgsqlCommand")]
        [Arguments("ProfiledDbCommand")]
        [Arguments("ExampleCommand")]
        [Arguments("Example")]
        [Arguments("Command")]
        [Arguments("Command")]
        public string OldGetDbType(string commandTypeName)
        {
            return ScopeFactory.OldGetDbType(commandTypeName);
        }

        [Benchmark]
        [Arguments("System.Data.SqlClient", "SqlCommand")]
        [Arguments("MySql.Data.MySqlClient", "MySqlCommand")]
        [Arguments("Npgsql", "NpgsqlCommand")]
        [Arguments("", "ProfiledDbCommand")]
        [Arguments("", "ExampleCommand")]
        [Arguments("", "Example")]
        [Arguments("", "Command")]
        [Arguments("Custom.DB", "Command")]
        public string GetDbType(string namespaceName, string commandTypeName)
        {
            return ScopeFactory.GetDbType(namespaceName, commandTypeName);
        }
```

Reply:
Cool, thanks! 😄 

Sorry for the nitpicking, I'm doing this because the method is in the hotpath...

The new code creates a [bigger codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHufv6nySAQAYGAcRgYAIsAAqATwAOMABTFyogHbZ8MXIuxgYAOR0w0IhpHz5smgCYLlp3QEouvHp2qffViJq4GJbWtg4AygCuAGbRAJYIDAC8DABELBA2dvapANw0Hn48AG7YUAywuJEANsFJhUWeodlOJmYMuADucRhgABYNjV6DQ56p4QCO1RlZDqnJAHxpuFNwuDBQxRupaCOjPKnGigDmK9OZYTmLaYoQQceVO3v7aQCy8pPns1dJS6n48jOTx8L184ymPRgM0u81+ywhGBgwNBYPCAEUADKQ6HZWF/M6Q5Eo3ipACSmkRUCMigw2GA1RgMhxc2umhq1V2IOJ3FSAAUoBB4gz7EyLrjWezOdyeAB9BidPowTQMbS6fSGNq6AB0GKVxwwfWSKVEADITf5vq0XDAjRbLlFYglrqlmnMpdKGHKFUqVWZ1UZrVqMhTsHFAioAORaiOuBhmu0tJSam1JFKu+wO+KJX7PYmqvQGANmLVRYDqYQqfP+5M6vAYcn2GAIADy0Uj0dcri1sggGIgnQ25NKUDidgwKlc7ulXsVyvTVvaqYTERiWcWuZRVcLNZ7fYHUCHZVHFInU+5M5986TgYAog5cAB1Hp9FTpzMJWM5rke7hX5zF0tyxUYQLD/GtdU0fVDTgZcM1XBIdT1A0u13ftB00Ydj3HScNyGOU4TAwNUP3Q8RzHCdcM8ABffJqEo4gAHYKj0GoMFoooqJoKigA=) with 180 instructions, that in itself is not a cause for concern, but I would like to try another approach (**of course we need to benchmark it**). And the reason is to check if by removing `ComputeStringHash` we can perform faster. The current code computes a hash before the Equals comparison (normal when a switch contains a lot of cases), so for each case we have a `Hash() + Jump + Equals` and calculating a hash is fast but not free.

The [following code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHufv6nySAQAYGAcRgYAIsAAqATwAOMUgApi5UQDts+GLkXYwMAHK6YaEQ0j582LQBMFys3oCUXXj07UvfhgD0AQwAYgCWULgYDADuMAwYUPKxYRgAFglp8fgQUdYQthBauAxF+gB0DACCxChhWgDmmfEsBYoArhgwAMqJ9Q0AEnhpnv48uDGpYBmqNnaOzqbmHr5j3qNrvGB48QBE3QCOADattvYOuyAbm37EAOwMu7jHcLgwUABu77sA3Nc32zejxMigazxOBXmFyuqxu/nuj0UuQwDVguF+/02gL2AFl5IcIWdHJdMXCmA9dvh5OCMbCydjHgB5KBGI4wU5Qkl0sk8BG7aCsmC0nncBn7Y6pdmQ85ckXWHaPboARQAMpKOTKYXLeHzwZLhTcAL6k3gmnhBBgASQAZrF4tgHA5mkDsSUMBBmgxFLAPmEIO0ShMphl0vEAFKW2QpI5HBjYRSKI7JUOsNqdHp9RpDXAZSAOGANGBaM3cIMYaYMWbShZKJbuDglrxiy1aLpQYyKDDYYBsmQa4lakVigAKUAg1rCbIcfer0MbYwtDiKAHJomBYNgugwDPYStboM6YPP4Q8tO0Yz9tTxG8buf5GxaTBAYp72lo5kV5W9cI3IMVohoohzOc3TtNaE4IAwAC8jz9hcfzUI2YS2lWRJOLWrjxFBMHAY4oHgWECBuAwjY+HKyGVjoeg7sYmHlCqRYNOk0EwcIxHHtwZFXuSjy4fBHEMLecoCfU0RHHgGAtvmkEwVR+iGLR5j0RJUkwAgjLWqoy7lMubgIeRKHiVEqmQQAfAwcDkOxd6bFxV4InJNF1jA5SgcAgGqEZkmOGpbjlLIEAqs+7wth82BQGE9gYKoekCUJIoCQ55hOXRAVBXEUCheFkWtjF+lGo2SEoXxix0QAoo4uAAOqpGkqFQvhEFuCscJ2WSCIlRhSluR5wiWJ1LhKQxjTMXA+RoY1hH0Yx6R+WlwWZVoYURVFeU3oVNleB1s6lUp80ZVlK25bFm23oaQA===) is bigger and more verbose in C# code, but has an smaller codegen (148 instructions):

```csharp
public static string GetDbType2(string namespaceName, string commandTypeName)
{
    // First we try with the most commons ones. Avoiding the ComputeStringHash
    switch (commandTypeName)
    {
        case "SqlCommand":
            return "sql-server";
        case "NpgsqlCommand":
            return "postgres";
        case "MySqlCommand":
            return "mysql";
        case "OracleCommand":
            return "oracle";
        case "SqliteCommand":
        case "SQLiteCommand":
            return "sqlite";
    }
    
    // If we add these cases to the previous switch the JIT will apply the ComputeStringHash codegen
    switch (commandTypeName) {
        case "InterceptableDbCommand":
        case "ProfiledDbCommand":
            // don't create spans for these
            return null;                    
    }
    
    // Now the uncommon cases
    const string commandSuffix = "Command";

    if (commandTypeName == commandSuffix) 
    {
        if (namespaceName.Length == 0) 
        {
            return "command";
        }
        
        int lastIndex = namespaceName.LastIndexOf('.');
        if (lastIndex > -1) 
        {
            return namespaceName.Substring(lastIndex).ToLowerInvariant();
        }
        
        return namespaceName.ToLowerInvariant();
    }
    
    if (commandTypeName.EndsWith(commandSuffix))
    {
        return commandTypeName.Substring(0, commandTypeName.Length - commandSuffix.Length).ToLowerInvariant();
    }
    
    return commandTypeName.ToLowerInvariant();
}
```

With this form, we are able to remove the `ComputeStringHash` and for the common cases (eg: SqlCommand, NpgsqlCommand) it should perform better.

I could be totally wrong with this, that's why we need to benchmark it.

Because you already have the BenchmarkDotnet project, can you try this algorithm as another BenchmarkDotNet case? :)

Reply:
So, I benchmarked all three methods together, here are the numbers:

|         Method |        namespaceName |   commandTypeName |       Mean |     Error |    StdDev |     Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |--------------------- |------------------ |-----------:|----------:|----------:|-----------:|-------:|------:|------:|----------:|
|  OldGetDbType |                    ? |        SqlCommand |   8.996 ns | 0.0462 ns | 0.0386 ns |   8.979 ns |      - |     - |     - |         - |
|     GetDbType | Syste(...)lient [21] |        SqlCommand |  10.081 ns | 0.0967 ns | 0.0755 ns |  10.072 ns |      - |     - |     - |         - |
| TonyGetDbType | Syste(...)lient [21] |        SqlCommand |   8.213 ns | 0.3600 ns | 1.0387 ns |   8.593 ns |      - |     - |     - |         - |
|  OldGetDbType |                    ? |      MySqlCommand |  10.280 ns | 0.1214 ns | 0.1077 ns |  10.265 ns |      - |     - |     - |         - |
|     GetDbType | MySql(...)lient [22] |      MySqlCommand |  10.765 ns | 0.0766 ns | 0.0640 ns |  10.756 ns |      - |     - |     - |         - |
| TonyGetDbType | MySql(...)lient [22] |      MySqlCommand |   7.085 ns | 0.1809 ns | 0.1692 ns |   7.041 ns |      - |     - |     - |         - |
|  OldGetDbType |                    ? |     NpgsqlCommand |   9.635 ns | 0.0343 ns | 0.0268 ns |   9.629 ns |      - |     - |     - |         - |
|     GetDbType |               Npgsql |     NpgsqlCommand |  11.115 ns | 0.0753 ns | 0.0628 ns |  11.098 ns |      - |     - |     - |         - |
| TonyGetDbType |               Npgsql |     NpgsqlCommand |   5.597 ns | 0.0654 ns | 0.0546 ns |   5.602 ns |      - |     - |     - |         - |

It seems TonyGetDbType is better for the common cases, but the same or worse for the slower methods. If the objective is getting rid of the call `ComputeStringHash` then I had the idea, this might be possible with a mixed apporach. This snip doesn't seem to need `ComputeStringHash` but still uses a more modern syntax:

```
        public static string MixedGetDbType(string namespaceName, string commandTypeName)
        {
            const string commandSuffix = "Command";

            var result =
                commandTypeName switch
                {
                    "SqlCommand" => "sql-server",
                    "NpgsqlCommand" => "postgres",
                    "MySqlCommand" => "mysql",
                    "SqliteCommand" => "sqlite",
                    "SQLiteCommand" => "sqlite",
                    _ => null,
                };

            if (result != null || result == "InterceptableDbCommand" || result == "ProfiledDbCommand")
            {
                return result;
            }

            return
                commandTypeName switch
                {
                    _ when namespaceName.Length == 0 && commandTypeName == commandSuffix => "command",
                    _ when namespaceName.Contains('.') && commandTypeName == commandSuffix =>
                        namespaceName.Substring(namespaceName.LastIndexOf('.')).ToLowerInvariant(),
                    _ when commandTypeName == commandSuffix =>
                        namespaceName.ToLowerInvariant(),
                    _ when commandTypeName.EndsWith(commandSuffix) =>
                        commandTypeName.Substring(0, commandTypeName.Length - commandSuffix.Length).ToLowerInvariant(),
                    _ => commandTypeName.ToLowerInvariant()
                };
        }

```

Here are the numbers:

|         Method |        namespaceName |   commandTypeName |       Mean |     Error |    StdDev |     Median |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |--------------------- |------------------ |-----------:|----------:|----------:|-----------:|-------:|------:|------:|----------:|
|   OldGetDbType |                    ? |           Command |  16.586 ns | 0.2255 ns | 0.1883 ns |  16.614 ns |      - |     - |     - |         - |
|      GetDbType |                      |           Command |  11.740 ns | 0.1858 ns | 0.1552 ns |  11.688 ns |      - |     - |     - |         - |
|  TonyGetDbType |                      |           Command |  20.234 ns | 0.1540 ns | 0.1286 ns |  20.177 ns |      - |     - |     - |         - |
| MixedGetDbType |                      |           Command |  23.969 ns | 1.3773 ns | 3.8620 ns |  22.571 ns |      - |     - |     - |         - |
|   OldGetDbType |                    ? |           Example | 115.235 ns | 2.3236 ns | 3.4060 ns | 115.476 ns | 0.0004 |     - |     - |      40 B |
|      GetDbType |                      |           Example | 125.505 ns | 1.9138 ns | 1.7902 ns | 124.983 ns | 0.0002 |     - |     - |      40 B |
|  TonyGetDbType |                      |           Example | 123.224 ns | 1.4947 ns | 1.3250 ns | 123.380 ns | 0.0002 |     - |     - |      40 B |
| MixedGetDbType |                      |           Example | 126.094 ns | 2.0159 ns | 1.7870 ns | 125.657 ns | 0.0002 |     - |     - |      40 B |
|   OldGetDbType |                    ? |    ExampleCommand | 142.110 ns | 3.4806 ns | 9.7600 ns | 140.215 ns | 0.0007 |     - |     - |      80 B |
|      GetDbType |                      |    ExampleCommand | 138.620 ns | 2.3756 ns | 1.9838 ns | 138.667 ns | 0.0007 |     - |     - |      80 B |
|  TonyGetDbType |                      |    ExampleCommand | 136.636 ns | 2.2792 ns | 2.2385 ns | 135.975 ns | 0.0007 |     - |     - |      80 B |
| MixedGetDbType |                      |    ExampleCommand | 140.400 ns | 1.8541 ns | 1.7343 ns | 140.636 ns | 0.0007 |     - |     - |      80 B |
|   OldGetDbType |                    ? | ProfiledDbCommand |  12.432 ns | 0.1877 ns | 0.1664 ns |  12.407 ns |      - |     - |     - |         - |
|      GetDbType |                      | ProfiledDbCommand |  13.779 ns | 0.2959 ns | 0.2768 ns |  13.793 ns |      - |     - |     - |         - |
|  TonyGetDbType |                      | ProfiledDbCommand |  19.760 ns | 0.3188 ns | 0.2662 ns |  19.839 ns |      - |     - |     - |         - |
| MixedGetDbType |                      | ProfiledDbCommand | 163.994 ns | 1.6919 ns | 1.4128 ns | 163.796 ns | 0.0010 |     - |     - |      96 B |
|   OldGetDbType |                    ? |           Command |  16.071 ns | 0.3000 ns | 0.2660 ns |  16.004 ns |      - |     - |     - |         - |
|      GetDbType |            Custom.DB |           Command |  52.357 ns | 1.1184 ns | 2.1006 ns |  51.976 ns | 0.0007 |     - |     - |      64 B |
|  TonyGetDbType |            Custom.DB |           Command |  62.060 ns | 0.9943 ns | 0.8814 ns |  61.656 ns | 0.0007 |     - |     - |      64 B |
| MixedGetDbType |            Custom.DB |           Command |  65.835 ns | 0.5969 ns | 0.5291 ns |  65.879 ns | 0.0007 |     - |     - |      64 B |
|   OldGetDbType |                    ? |        SqlCommand |   9.620 ns | 0.2568 ns | 0.2402 ns |   9.568 ns |      - |     - |     - |         - |
|      GetDbType | Syste(...)lient [21] |        SqlCommand |  10.830 ns | 0.2628 ns | 0.3769 ns |  10.859 ns |      - |     - |     - |         - |
|  TonyGetDbType | Syste(...)lient [21] |        SqlCommand |   3.423 ns | 0.1470 ns | 0.2613 ns |   3.400 ns |      - |     - |     - |         - |
| MixedGetDbType | Syste(...)lient [21] |        SqlCommand |   4.608 ns | 0.1611 ns | 0.2601 ns |   4.506 ns |      - |     - |     - |         - |
|   OldGetDbType |                    ? |      MySqlCommand |  10.561 ns | 0.1827 ns | 0.1709 ns |  10.573 ns |      - |     - |     - |         - |
|      GetDbType | MySql(...)lient [22] |      MySqlCommand |  10.961 ns | 0.2897 ns | 0.5298 ns |  10.754 ns |      - |     - |     - |         - |
|  TonyGetDbType | MySql(...)lient [22] |      MySqlCommand |   7.326 ns | 0.1225 ns | 0.1086 ns |   7.366 ns |      - |     - |     - |         - |
| MixedGetDbType | MySql(...)lient [22] |      MySqlCommand |   9.049 ns | 0.3187 ns | 0.9145 ns |   9.064 ns |      - |     - |     - |         - |
|   OldGetDbType |                    ? |     NpgsqlCommand |  10.856 ns | 0.1980 ns | 0.1756 ns |  10.898 ns |      - |     - |     - |         - |
|      GetDbType |               Npgsql |     NpgsqlCommand |   9.957 ns | 0.1305 ns | 0.1221 ns |  10.001 ns |      - |     - |     - |         - |
|  TonyGetDbType |               Npgsql |     NpgsqlCommand |   6.220 ns | 0.1684 ns | 0.1576 ns |   6.241 ns |      - |     - |     - |         - |
| MixedGetDbType |               Npgsql |     NpgsqlCommand |   6.489 ns | 0.2090 ns | 0.2488 ns |   6.338 ns |      - |     - |     - |         - |

I see there some variation from the last test, and the number for the MixedGetDbType / ProfiledDbCommand case looks wrong, probably something happening on my machine at the time. However, MixedGetDbType seems to preform as well as TonyGetDbTypes for the common cases and still does reasonably in the other cases, so shall we go with that case?

Reply:
> It seems TonyGetDbType is better for the common cases, but the same or worse for the slower methods. 

Yes, that's expected, having a lot `Equals` after each jump behave worse on latest/bottom cases. But better in the initial cases. So this kind of optimization works for us due the common cases are on the top of the method.

> I see there some variation from the last test, and the number for the MixedGetDbType / ProfiledDbCommand case looks wrong, probably something happening on my machine at the time.

Yes, that result was weird, I'm not sure what happened there, is also allocating....

> However, MixedGetDbType seems to preform as well as TonyGetDbTypes for the common cases and still does reasonably in the other cases, so shall we go with that case?

Sure, if you want to use the new switch syntax go for it, the results between both are negligible.

Reply:
```csharp
if (result != null || result == "InterceptableDbCommand" || result == "ProfiledDbCommand")
```
This should be:
```csharp
if (result != null || commandTypeName == "InterceptableDbCommand" || commandTypeName == "ProfiledDbCommand")
```
It explains the inconsistency for the ProfiledDbCommand case.

Reply:
D'oh. Ironically my unit test caught this and another small bug, but I forgot to check them before I pushed :(